lib/sources.nix

```diff
@@ -26,14 +26,32 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = src: cleanSourceWith { filter = cleanSourceFilter; inherit src; };
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  cleanSourceWith = { filter, src }:
```

Comment:
An additional/alternative design that I've used is to write `composeFilter` and just use that to build up bigger filters.